src/EditorFeatures/CSharpTest/ReassignedVariable/CSharpReassignedVariableTests.cs

```diff
@@ -0,0 +1,853 @@
+    {
+        [|p|] = 0;
+        [|p|] = 1;
+        Console.WriteLine([|p|]);
+    }
+}");
+        }
+
+        [Fact]
+        public async Task TestOutParameterReassignmentOneWrites()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    void M(out int p)
+    {
+        p = ref p;
+        Console.WriteLine(p);
+    }
+}");
+        }
+
+        [Fact]
+        public async Task AssignmentThroughRefParameter()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    void M(ref int [|p|])
+    {
+        [|p|] = 0;
+        [|p|] = 1;
+        Console.WriteLine([|p|]);
+    }
+}");
+        }
+
+        [Fact]
+        public async Task TestRefParameterReassignment()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    void M(ref int [|p|])
+    {
+        [|p|] = ref [|p|];
+        Console.WriteLine([|p|]);
+    }
+}");
+        }
+
+        [Fact]
+        public async Task AssignmentThroughRefLocal()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    void M(ref int [|p|])
+    {
+        ref var [|local|] = ref [|p|];
+        [|local|] = 0;
+        [|local|] = 1;
+        Console.WriteLine([|local|]);
+    }
+}");
+        }
+
+        [Fact]
+        public async Task TestRefLocalReassignment()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    void M(ref int [|p|])
+    {
+        ref var [|local|] = ref [|p|];
+        [|local|] = ref [|p|];
+        Console.WriteLine([|local|]);
+    }
+}");
+        }
+
+        [Fact]
+        public async Task AssignmentThroughPointerIsNotAssignmentOfTheVariableItself()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    unsafe void M(int* p)
+    {
+        *p = 4;
+        Console.WriteLine((IntPtr)p);
+    }
+}");
+        }
+
+        [Fact]
+        public async Task TestPointerVariableReassignment()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    unsafe void M(int* [|p|])
+    {
+        [|p|] = null;
+        Console.WriteLine((IntPtr)[|p|]);
+    }
+}");
+        }
+
+        [Fact]
+        public async Task TestRefParameterCausingPossibleReassignment()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    void M()
+    {
+        int [|p|] = 0;
+        M2(ref [|p|]);
+        Console.WriteLine([|p|]);
+    }
+
+    void M2(ref int p) { }
+}");
+        }
+
+        [Fact]
+        public async Task TestRefParameterWithoutReassignment()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    void M()
+    {
+        int p;
+        M2(ref p);
+        Console.WriteLine(p);
+    }
+
+    void M2(ref int p) { }
+}");
+        }
+
+        [Fact]
+        public async Task TestRefLocalCausingPossibleReassignment()
+        {
+            await TestAsync(
+@"
+using System;
+class C
+{
+    void M()
+    {
+        int [|p|] = 0;
+        ref int refP = ref [|p|];
```

Comment:
💡 Add a matching test for `readonly ref int refP`, where `p` is not detected as written.

Reply:
done.